### PR TITLE
Add machine learning and xgboost utils

### PR DESCRIPTION
**Pull request by TedBrookings**
Addded abstract class MachineLearningUtils to provide an interface and
handle common tasks. These include loading data, splitting data into
training and test sets, cross-validation, and optimizing classifier
hyperparameters.

Also added XGBoostUtils which provides a concrete implemention of
MachineLearningUtils (by wrapping xgboost4j) and serves as an example
of how to provide access to a 3rd-party machine learning library.

Finally, added an example tool: ExampleTrainXGBoostClassifier. This
demonstrates a typical training use case of loading data, training a
classifier, assessing accuracy, and saving the classifier. It also
demonstrates a typical filtering use case of loading a saved classifer,
and using it to calculate probabilities or class labels.

This is working towards issue 4922 by providing the tools necessary to
train classifiers in general, but does not provide tools to train a
BreakpointEvidence filter, so does not resolve it. Additionally, this
framework should eventually be extended to provide a bayesian
hyperparameter optimizer.

One outstanding problem with these changes is that xgboost4j threading
does not appear to work on OSX, resulting in slower training. However,
it does work on linux.

### CODE REVIEW

**Comment by droazen** [comment]
@cmnbroad Could you have a look at this one as well?

**Comment by TedBrookings** [comment]
Sorry to make this PR and run, I'll be on vacation until 9/12, but I'll take a look at the comments when I get back.

**Comment by SHuang-Broad** [comment]
In addition, I haven't been very through about how the random number generator is interacting with the multi-threading part, so just to be cautious: https://stackoverflow.com/questions/22313552/contention-in-concurrent-use-of-java-util-random
